text/0015-Service-Reference-Retrieval.md

```diff
@@ -0,0 +1,233 @@
+This document discusses need for such API in the particular use case of declarative services (DS) and its possible design/solution.
+
+## Motivation
+
+Developers using DS rely on DS's SCR (service component runtime) to inject references of services statically at construction or dynamically using the <code>bind</code> and <code>unbind</code> methods. These <code>ServiceObject</code>s can then be used by the client. However, clients often use properties in the Service to store metadata relevant to that service. Those properties, however, are not available to the <code>ServiceObject</code> itself, only its reference.
+
+Given this it seems appropriate to have a new method that the client can use to translate a <code>ServiceObject</code> into its <code>ServiceReference</code>.
+
+## Requirements
+
+This API must:
+
+1. take in a <code>ServiceObject\<void\></code> and return the <code>ServiceReference\<void\></code> for that service
+2. take in a <code>ServiceObject\<T\></code> and return the <code>ServiceReference\<T\></code> for that service where <code>T</code> is any class interface
+3. not expose internal <code>ServiceObject</code> implementation details to the user
+
+## Detailed design
+
+<code>ServiceReferenceFromService</code> API will be declared in the file <i>framework/include/cppmicroservices/ServiceReference.h</i> in the namespace <code><bold>cppmicroservices::</bold></code> as below:
+
+```c++
+
+/**
+* \ingroup MicroServices
+* \ingroup gr_servicereference
+*
+* A method to retrieve a <code>ServiceObject</code>'s original <code>ServiceReference<void></code>
+*
+*/
+US_Framework_EXPORT ServiceReferenceU ServiceReferenceFromService(std::shared_ptr<void> const& s);
+
+/**
+* \ingroup MicroServices
+* \ingroup gr_servicereference
+*
+* A method to retrieve a <code>ServiceObject<T></code>'s original <code>ServiceReference<U></code>
+*
+* @tparam T The class type of the <code>ServiceObject</code>
+* @tparam U The class type of the <code>ServiceReference</code>. Defaults to <code>T</code> if not specified.
+*/
+template <typename T, typename U = T>
+ServiceReference<U>
+ServiceReferenceFromService(std::shared_ptr<T> const& s)
+{
+    return ServiceReference<U>(ServiceReferenceFromService(std::static_pointer_cast<void>(s)));
+}
+
+```
+
+This API will take the <code>ServiceObject</code> and first verify that it was a <code>ServiceObject</code> created by the CppMicroServices framework.
+
+If it is valid, it will return the original <code>ServiceReference</code> for that <code>ServiceObject</code>.
+
+A typical usage workflow could be as below, in the static constructor of a service <code>ServiceImpl</code> which depends on a service of type <code>ServiceInterface2</code>.
+
+```c++
+ServiceImpl::ServiceImpl(std::shared_ptr<ServiceInterface2> const& dep) {
+    // get the reference
+    ServiceReference<ServiceInterface2> depRef = cppmicroservices::ServiceReferenceFromService<ServiceInterface>(dep);
```

Comment:
the template param could be omitted (it should be deduced by the compiler I believe?), or if specified, it should be `ServiceInterface2` ?

Reply:
Yes that is correct. I was confused about this. Is the purpose of the non-templated one only so that something can be written in the actual cpp files as opposed to the headers as the templated one must be fully exposed as to implementation?